packages/adapter-components/src/client/rate_limiter.ts

```diff
@@ -0,0 +1,291 @@
+    }
+    this.queue = this.options.useBottleneck
+      ? new Bottleneck({
+          maxConcurrent: this.options.maxConcurrentCalls === Infinity ? null : this.options.maxConcurrentCalls,
+          reservoir: this.options.maxCallsPerInterval === Infinity ? null : this.options.maxCallsPerInterval,
+          reservoirRefreshInterval: this.options.intervalLengthMS === 0 ? null : this.options.intervalLengthMS,
+          reservoirRefreshAmount:
+            this.options.maxCallsPerInterval === Infinity ? null : this.options.maxCallsPerInterval,
+        })
+      : new PQueue({
+          concurrency: this.options.maxConcurrentCalls,
+          intervalCap: this.options.maxCallsPerInterval,
+          interval: this.options.intervalLengthMS,
+          carryoverConcurrencyCount: this.options.carryRunningCallsOver,
+          autoStart: !this.options.startPaused,
+        })
+
+    this.counters = {
+      total: 0,
+      running: 0,
+      pending: 0,
+      done: 0,
+    }
+    this.prevInvocationTime = Date.now()
+  }
+
+  /**
+   * Returns a copy of the current configuration options of the rate limiter.
+   * @returns The current configuration options.
+   */
+  getOptions(): RateLimiterOptions {
+    return { ...this.options }
+  }
+
+  /**
+   * Calculates the delay required to maintain the configured delay between task invocations.
+   * It ensures that tasks are spaced out by at least the delay specified in the options.
+   * @returns The calculated delay in milliseconds.
+   */
+  private nextDelay(): number {
+    const currTime = Date.now()
+    const timeSinceLastTask = currTime - this.prevInvocationTime
+    const delay = Math.max(0, this.options.delayMS - timeSinceLastTask)
+    this.prevInvocationTime = currTime + delay
+
+    return delay
+  }
+
+  /**
+   * Wraps a given task with a delay, ensuring that there is a pause between task invocations.
+   * If a delay is specified in the options, it waits for the calculated delay before executing the task.
+   * @param task The task to be wrapped and delayed.
+   * @returns A function that returns a promise resolving to the result of the task.
+   */
+  private getDelayedTask<T>(task: () => Promise<T>): () => Promise<T> {
+    const delayedTask = async (): Promise<T> => {
+      if (this.options.delayMS) {
+        await new Promise(resolve => setTimeout(resolve, this.nextDelay()))
+      }
+      return task()
+    }
+    return delayedTask
+  }
+
+  /**
+   * Wraps a given task with counters bookkeeping.
+   * @param task The task to wrapped with bookkeeping.
+   * @returns A function that returns a promise resolving to the result of the task
+   */
+  private wrapWithBookKeeping<T>(task: () => Promise<T>): () => Promise<T> {
+    return async (): Promise<T> => {
+      this.counters.pending -= 1
+      this.counters.running += 1
+      const res = await task()
+      this.counters.running -= 1
+      this.counters.done += 1
+      return res
```

Comment:
WDYT about this?
```suggestion
      try {
        const res = await task()
        this.counters.done += 1
        return res
      } catch (e) {
        this.counters.failed += 1
        throw e
      } finally {
        this.counters.running -= 1
      } 
```

or 

```suggestion
      try {
        return await task()
      } finally {
        this.counters.running -= 1
        this.counters.done += 1
      } 
```

Reply:
Nice idea :) I already wrote this logic in the retry mechanism PR.